csharp/src/Microsoft.ML.OnnxRuntime/NativeOnnxTensorMemory.shared.cs

```diff
@@ -248,7 +248,7 @@ public override MemoryHandle Pin(int elementIndex = 0)
                 {
                     throw new ArgumentOutOfRangeException(nameof(elementIndex));
                 }
-                return new MemoryHandle((void*)((int)_dataBufferPointer + elementIndex * ElementWidth)); //could not use Unsafe.Add
+                return new MemoryHandle(IntPtr.Add(_dataBufferPointer, elementIndex * ElementWidth).ToPointer());
```

Comment:
>elementIndex [](http://example.com/codeflow?start=71&length=12)

I feel we need to promote this to 64-bit first so we do not get overflow here even before adding.

Reply:
updated

Reply:
On 32bit platform, if the value of `_dataBufferPointer.ToInt64() + (long)elementIndex * ElementWidth` is in the range of Int32, this works. Otherwise, `new IntPtr(long)` will not throw `Overflowexception`. I think this is behavior we want.